File: 序号/857.雇佣K名工人的最低成本.kt
fun mincostToHireWorkers(quality: IntArray, wage: IntArray, k: Int): Double {
  // 因为同工同酬，按工作质量分发工资，所以一个人的工作质量占总工作质量比例 等于 他的工资占全部工资的比例
  // 总支付工资为 totalMoney，总工作质量为totalQuality，应该达成 totalMoney*（qualityI/totalQuality)>=wageI
  // totalMoney>= totalQuality*wageI/qualityI ,这意味着选一拨人，他们的工作质量是固定的，所要支付的工资 大于 wageI/qualityI
  // 按照wageI/qualityI为性价比排序，要选k个人，就从第k个人开始计算，计算wage.SIZE-K次需要支付的总金额，即可得出最小支付额度
  var sortIndexArr = List(wage.size) { it }
  sortIndexArr=sortIndexArr.sortedWith(object : Comparator<Int> {
    override fun compare(o1: Int, o2: Int): Int {
      return wage[o1] * quality[o2] - wage[o2] * quality[o1]
    }
  })
  
  var totalQ = 0.0
  val pq = PriorityQueue<Int>(object : Comparator<Int> {
    override fun compare(o1: Int, o2: Int): Int {
      return o2-o1
    }
  }) // pq 默认小顶堆，题设需要最小金钱，金钱与质量成正比线性关系，需要将最大质量移除
  for (i in 0 until k - 1) {
    val q = quality[sortIndexArr[i]]
    totalQ += q
    pq.add(q)
  }
  var ans = Double.MAX_VALUE
  for (i in k - 1 until wage.size) {
    val q = quality[sortIndexArr[i]]
    val w = wage[sortIndexArr[i]]
    totalQ += q
    pq.offer(q)
    ans = minOf(ans,  w.toDouble() / q*totalQ)
    totalQ -= pq.poll()
  }
  return ans
}

File: 序号/857. 雇佣 K 名工人的最低成本.py
class Solution:
    def mincostToHireWorkers(self, quality: List[int], wage: List[int], k: int) -> float:
        # 按工作质量发工资，且必须满足最低要求
        # 让某个人以最低工资发放，其他人
        # q=10，5   w = 100，20   0号发100，1号发50， 0号资质比是10，1号是4，选0作为基准，1号肯定能满足要求
        # 按工资/质量的比值排序，选前k个人求总工资。 往后遍历检查，后面的比值虽然会升高，但质量降低也有可能出现总值更低的答案
        p = sorted(zip(quality,wage),key=lambda x:x[1]/x[0])
        hq = [-q for q,_ in p[:k]] # 质量的大顶堆
        heapify(hq)
        sm = -sum(hq)
        ans = sm*p[k-1][1]/p[k-1][0]
        for q,w in p[k:]:
            if q < -hq[0]:
                topQ = -heappop(hq)
                heappush(hq,-q)
                sm = sm-topQ+q
                ans=min(ans,sm*w/q)
        return ans